File: data/derived_cot/rq2_eval/translation/qwen/_converted_code/java_to_py/DecryptionUtils.py
class DecryptionUtils:
    def __init__(self, key):
        self.key = key

    def caesar_decipher(self, ciphertext, shift):
        shift %= 26
        plaintext = []
        for char in ciphertext:
            if char.isalpha():
                ascii_offset = ord('A') if char.isupper() else ord('a')
                shifted_char = chr(((ord(char) - ascii_offset - shift) % 26) + ascii_offset)
                plaintext.append(shifted_char)
            else:
                plaintext.append(char)
        return ''.join(plaintext)

    def vigenere_decipher(self, ciphertext):
        decrypted_text = []
        key_index = 0
        for char in ciphertext:
            if char.isalpha():
                shift = ord(self.key[key_index % len(self.key)].lower()) - ord('a')
                base = ord('A') if char.isupper() else ord('a')
                decrypted_char = chr(((ord(char) - base - shift) % 26) + base)
                decrypted_text.append(decrypted_char)
                key_index += 1
            else:
                decrypted_text.append(char)
        return ''.join(decrypted_text)

    def rail_fence_decipher(self, encrypted_text, rails):
        if rails == 0:
            raise ValueError("Number of rails must be at least 1")
        if encrypted_text == "":
            return ""
        
        fence = [['' for _ in range(len(encrypted_text))] for _ in range(rails)]
        direction = -1
        row, col = 0, 0
        
        for char in encrypted_text:
            fence[row][col] = '*'
            if row == 0 or row == rails - 1:
                direction *= -1
            row += direction
            col += 1
        
        index = 0
        for i in range(rails):
            for j in range(len(encrypted_text)):
                if fence[i][j] == '*':
                    fence[i][j] = encrypted_text[index]
                    index += 1
        
        plaintext = []
        direction = -1
        row, col = 0, 0
        for _ in range(len(encrypted_text)):
            plaintext.append(fence[row][col])
            if row == 0 or row == rails - 1:
                direction *= -1
            row += direction
            col += 1
        
        return ''.join(plaintext)

File: data/derived_cot/rq1_traces/translation/r1/java_to_cpp/code_output/DecryptionUtils.cpp
#include <vector>
#include <string>
#include <cctype>

class DecryptionUtils {
private:
    std::string key;

public:
    DecryptionUtils(const std::string& key) : key(key) {}

    std::string caesarDecipher(const std::string& ciphertext, int shift) {
        std::string plaintext;
        shift = shift % 26;
        if (shift < 0) {
            shift += 26;
        }
        for (char ch : ciphertext) {
            if (std::isalpha(ch)) {
                char base = std::isupper(ch) ? 'A' : 'a';
                char shiftedChar = (ch - base - shift + 26) % 26 + base;
                plaintext += shiftedChar;
            } else {
                plaintext += ch;
            }
        }
        return plaintext;
    }

    std::string vigenereDecipher(const std::string& ciphertext) {
        std::string decryptedText;
        int keyIndex = 0;
        for (char ch : ciphertext) {
            if (std::isalpha(ch)) {
                char keyChar = key[keyIndex % key.length()];
                if (std::isupper(keyChar)) {
                    keyChar = std::tolower(keyChar);
                }
                int shift = keyChar - 'a';
                char base = std::isupper(ch) ? 'A' : 'a';
                char decryptedChar = (ch - base - shift + 26) % 26 + base;
                decryptedText += decryptedChar;
                keyIndex++;
            } else {
                decryptedText += ch;
            }
        }
        return decryptedText;
    }

    std::string railFenceDecipher(const std::string& encryptedText, int rails) {
        if (rails <= 0) {
            return encryptedText;
        }
        int len = encryptedText.length();
        std::vector<std::vector<char>> fence(rails, std::vector<char>(len, '\0'));

        int direction = -1;
        int row = 0, col = 0;

        for (int i = 0; i < len; i++) {
            if (row == 0 || row == rails - 1) {
                direction = -direction;
            }
            fence[row][col] = '*';
            col++;
            row += direction;
        }

        int index = 0;
        for (int i = 0; i < rails; i++) {
            for (int j = 0; j < len; j++) {
                if (fence[i][j] == '*') {
                    fence[i][j] = encryptedText[index];
                    index++;
                }
            }
        }

        std::string plainText;
        direction = -1;
        row = 0;
        col = 0;

        for (int i = 0; i < len; i++) {
            if (row == 0 || row == rails - 1) {
                direction = -direction;
            }
            plainText += fence[row][col];
            col++;
            row += direction;
        }

        return plainText;
    }
};